In [1]:
import re
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
%matplotlib inline

In [3]:
master = pd.read_csv("data/Master.csv")
batting = pd.read_csv("data/Batting.csv")
salaries = pd.read_csv("data/Salaries.csv")
positions = pd.read_csv("data/Appearances.csv")
pitching = pd.read_csv("data/Pitching.csv")

In [4]:
part_master = master[["playerID", "nameFirst", "nameLast"]]
part_batting = batting[["playerID", "yearID", "teamID", "lgID", "G", "AB", "H", "BB", "HBP", "SF"]]
part_salaries = salaries[["playerID", "yearID", "teamID", "salary"]]
part_positions = positions[["playerID", "yearID", "teamID", "GS", "G_p", "G_c", "G_1b", "G_2b", "G_3b", "G_ss", "G_lf", "G_cf", "G_rf", "G_of", "G_dh"]]
part_pitching = pitching[["playerID", "yearID", "teamID", "W", "L", "GS", "ERA"]]

In [5]:
batting_2004 = part_batting[part_batting["yearID"] == 2004]
salaries_2004 = part_salaries[part_salaries["yearID"] == 2004]
positions_2004 = part_positions[part_positions["yearID"] == 2004]
pitching_2004 = part_pitching[part_pitching["yearID"] == 2004]

In [69]:
pitching_2004.sort("W", ascending=False).head()

playerID  yearID teamID   W   L  GS   ERA
36106  schilcu01    2004    BOS  21   6  32  3.26
36104  santajo01    2004    MIN  20   6  34  2.61
36019  oswalro01    2004    HOU  20  10  35  3.49
36107  schmija01    2004    SFN  18   7  32  3.20
35676  clemero02    2004    HOU  18   4  33  2.98

In [6]:
battpitch_2004 = pd.merge(batting_2004, pitching_2004, how="outer", on=("playerID", "teamID"))

In [7]:
battpitchpos_2004 = pd.merge(battpitch_2004, positions_2004, how="outer", on=("playerID", "teamID"))

In [8]:
battpitchpossal_2004 = pd.merge(battpitchpos_2004, salaries_2004, how="outer", on=("playerID", "teamID"))

In [10]:
mlb_2004 = pd.merge(battpitchpossal_2004, part_master, how="inner", on="playerID")

In [14]:
mlb_2004["OBP"] = (mlb_2004.H + mlb_2004.BB + mlb_2004.HBP) / (mlb_2004.AB + mlb_2004.BB + mlb_2004.HBP + mlb_2004.SF)

In [16]:
OBP_2004 = mlb_2004.OBP

In [18]:
avg_obp_2004 = OBP_2004.mean()

In [21]:
stdev_obp_2004 = OBP_2004.std()

In [24]:
mlb_2004["OBP_std"] = (mlb_2004.OBP - avg_obp_2004) / stdev_obp_2004

In [45]:
avg_sal_2004 = mlb_2004.salary.mean()
stdev_sal_2004 = mlb_2004.salary.std()

In [70]:
avg_sal_2004

2489998.6775000002

In [46]:
mlb_2004["salary_std"] = (mlb_2004.salary - avg_sal_2004) / stdev_sal_2004

In [47]:
mlb_2004["OBP_Sal"] = mlb_2004.OBP_std - mlb_2004.salary_std

In [48]:
mlb_2004.sort("OBP_Sal", ascending=False).head(20)

playerID  yearID_x teamID lgID    G   AB    H  BB  HBP  SF    ...     \
258   cottsne01      2004    CHA   AL   56    1    1   0    0   0    ...      
146   broweji01      2004    SFN   NL   89    2    1   1    0   0    ...      
352   eischjo01      2004    MON   NL   22    3    2   0    0   0    ...      
1207  torcato01      2004    SFN   NL   13    9    5   1    1   1    ...      
97    bentzch01      2004    MON   NL   36    2    1   0    0   0    ...      
675   knottga01      2004    DET   AL   36    3    1   1    0   0    ...      
34    ankieri01      2004    SLN   NL    5    1    0   1    0   0    ...      
922   olmedra01      2004    CIN   NL    8    1    0   1    0   0    ...      
457   gipsoch01      2004    TBA   AL    5    4    2   0    0   0    ...      
1131  simonja01      2004    SLN   NL   15    2    1   0    0   0    ...      
834   micelda01      2004    HOU   NL   74    2    1   0    0   0    ...      
1033   reedst01      2004    COL   NL   65    2    1   0    0   0    ...      
1209  torresa01      2004    PIT   NL   84    2    1   0    0   0    ...      
60     bardjo01      2004    CLE   AL    7   19    8   3    0   1    ...      
1111  schoesc01      2004    CHA   AL   20    2    1   0    0   0    ...      
1024  randost01      2004    ARI   NL   47   12    5   0    0   0    ...      
526   hafnetr01      2004    CLE   AL  140  482  150  68   17   6    ...      
911   obermwe01      2004    MIL   NL   26   39   15   1    0   0    ...      
417   frankwa01      2004    SFN   NL   43    3    1   1    0   1    ...      
541   hanseda01      2004    SEA   AL   57   78   22  18    0   1    ...      

      G_of  G_dh  yearID_y   salary  nameFirst     nameLast       OBP  \
258      0     0      2004   301000       Neal        Cotts  1.000000   
146      0     0      2004   662500        Jim       Brower  0.666667   
352      0     0      2004  1300000       Joey      Eischen  0.666667   
1207     0     0      2004   302000       Tony      Torcato  0.583333   
97       0     0      2004   300000       Chad        Bentz  0.500000   
675      0     0      2004   316000       Gary       Knotts  0.500000   
34       0     0      2004   320000       Rick       Ankiel  0.500000   
922      0     0      2004   322000        Ray       Olmedo  0.500000   
457      2     0      2004   350000    Charles       Gipson  0.500000   
1131     1     0      2004   350000      Jason  Simontacchi  0.500000   
834      0     0      2004   600000        Dan       Miceli  0.500000   
1033     0     0      2004   600000      Steve         Reed  0.500000   
1209     0     0      2004   775000    Salomon       Torres  0.500000   
60       0     0      2004   316700       Josh         Bard  0.478261   
1111     0     0      2004  1725000      Scott  Schoeneweis  0.500000   
1024     0     0      2004   322500      Steve     Randolph  0.416667   
526      0   121      2004   316300     Travis       Hafner  0.410122   
911      0     0      2004   314000        Wes  Obermueller  0.400000   
417      0     0      2004   360000      Wayne     Franklin  0.400000   
541      0     8      2004   750000       Dave       Hansen  0.412371   

       OBP_std  salary_std   OBP_Sal  
258   5.033774   -0.619846  5.653620  
146   2.806724   -0.517482  3.324206  
352   2.806724   -0.336965  3.143689  
1207  2.249961   -0.619563  2.869524  
97    1.693199   -0.620129  2.313328  
675   1.693199   -0.615599  2.308797  
34    1.693199   -0.614466  2.307665  
922   1.693199   -0.613900  2.307098  
457   1.693199   -0.605971  2.299170  
1131  1.693199   -0.605971  2.299170  
834   1.693199   -0.535180  2.228379  
1033  1.693199   -0.535180  2.228379  
1209  1.693199   -0.485626  2.178825  
60    1.547956   -0.615400  2.163357  
1111  1.693199   -0.216620  1.909819  
1024  1.136436   -0.613758  1.750194  
526   1.092711   -0.615514  1.708225  
911   1.025084   -0.616165  1.641249  
417   1.025084   -0.603139  1.628223  
541   1.107737   -0.492705  1.600

At this point, I have a nice dataframe called mlb_2004 that has batting, pitching, appearance, and salary stats from the 2004 MLB season. I've also added in: OBP (on base percentage), OBP_std (how many standard deviations above or below the mean OBP that player is), salary_std (how many standard deviations above or below the mean salary that player is), OBP_Sal (OBP_std minus salary_std).

This last number can help compare players based on their OBP relative to the average along with their salary relative to the average.

In [50]:
mlb_2004_pitchers = mlb_2004[mlb_2004["G_p"] > 0]

In [77]:
mlb_2004_pitchers = mlb_2004_pitchers[mlb_2004_pitchers.OBP.notnull()]

In [89]:
mlb_2014_pitchers[mlb_2014_pitchers["lgID"] == "NL"].sort("OBP_Sal", ascending=False).head(20)


playerID  yearID_x teamID lgID    G   AB   H  BB  HBP  SF    ...     \
146   broweji01      2004    SFN   NL   89    2   1   1    0   0    ...      
352   eischjo01      2004    MON   NL   22    3   2   0    0   0    ...      
97    bentzch01      2004    MON   NL   36    2   1   0    0   0    ...      
34    ankieri01      2004    SLN   NL    5    1   0   1    0   0    ...      
1131  simonja01      2004    SLN   NL   15    2   1   0    0   0    ...      
1033   reedst01      2004    COL   NL   65    2   1   0    0   0    ...      
834   micelda01      2004    HOU   NL   74    2   1   0    0   0    ...      
1209  torresa01      2004    PIT   NL   84    2   1   0    0   0    ...      
1024  randost01      2004    ARI   NL   47   12   5   0    0   0    ...      
911   obermwe01      2004    MIL   NL   26   39  15   1    0   0    ...      
417   frankwa01      2004    SFN   NL   43    3   1   1    0   1    ...      
1160  stantmi02      2004    NYN   NL   83    2   1   0    0   0    ...      
378   falkebr01      2004    LAN   NL    6    2   0   1    0   0    ...      
47    ayalalu01      2004    MON   NL   81    9   3   0    0   0    ...      
274    cruzju02      2004    ATL   NL   50    5   1   1    0   0    ...      
666   kiescbr01      2004    MIL   NL   77   63  17   5    0   0    ...      
1244  venturo01      2004    LAN   NL  102  152  37  22    0   1    ...      
337   duckwbr01      2004    HOU   NL   19    9   2   1    0   0    ...      
353   elartsc01      2004    COL   NL    8    9   2   1    0   0    ...      
1343  zeileto01      2004    NYN   NL  137  348  81  44    1   2    ...      

      G_of  G_dh  yearID_y   salary  nameFirst     nameLast       OBP  \
146      0     0      2004   662500        Jim       Brower  0.666667   
352      0     0      2004  1300000       Joey      Eischen  0.666667   
97       0     0      2004   300000       Chad        Bentz  0.500000   
34       0     0      2004   320000       Rick       Ankiel  0.500000   
1131     1     0      2004   350000      Jason  Simontacchi  0.500000   
1033     0     0      2004   600000      Steve         Reed  0.500000   
834      0     0      2004   600000        Dan       Miceli  0.500000   
1209     0     0      2004   775000    Salomon       Torres  0.500000   
1024     0     0      2004   322500      Steve     Randolph  0.416667   
911      0     0      2004   314000        Wes  Obermueller  0.400000   
417      0     0      2004   360000      Wayne     Franklin  0.400000   
1160     0     0      2004  3000000       Mike      Stanton  0.500000   
378      0     0      2004   302500      Brian   Falkenborg  0.333333   
47       0     0      2004   317500       Luis        Ayala  0.333333   
274      0     0      2004   370000       Juan         Cruz  0.333333   
666      0     0      2004   380000     Brooks   Kieschnick  0.323529   
1244     0     0      2004  1200000      Robin      Ventura  0.337143   
337      0     0      2004   390000    Brandon    Duckworth  0.300000   
353      0     0      2004   480000      Scott      Elarton  0.300000   
1343     0     0      2004  1000000       Todd        Zeile  0.318987   

       OBP_std  salary_std   OBP_Sal  
146   2.806724   -0.517482  3.324206  
352   2.806724   -0.336965  3.143689  
97    1.693199   -0.620129  2.313328  
34    1.693199   -0.614466  2.307665  
1131  1.693199   -0.605971  2.299170  
1033  1.693199   -0.535180  2.228379  
834   1.693199   -0.535180  2.228379  
1209  1.693199   -0.485626  2.178825  
1024  1.136436   -0.613758  1.750194  
911   1.025084   -0.616165  1.641249  
417   1.025084   -0.603139  1.628223  
1160  1.693199    0.144414  1.548785  
378   0.579674   -0.619421  1.199095  
47    0.579674   -0.615174  1.194847  
274   0.579674   -0.600308  1.179981  
666   0.514172   -0.597476  1.111648  
1244  0.605126   -0.365281  0.970407  
337   0.356969   -0.594644  0.951613  
353   0.356969   -0.569160  0.926128  
1343  0.483826   -0.421914  0.905740  

[20 rows x 36 

Johan Santana from Minnesota is a good candidate for starting pitcher. His record in 2004 was 20-6, with a 2.61 ERA, so he's an excellent pitcher. But he's also a decent hitter. His OBP that year was .375, which is not quite 1 full standard deviation above the mean (0.858 to be exact), and his salary was just below the mean (1.6 million to the avg of 2.5 million). So his OBP_Sal is 1.11, which is much better than any of the other pitchers around him based on total wins. Unfortunately, he's an AL pitcher, so he only had 8 ABs, which means his OBP might be an anomoly. So we'll restrict this to NL pitchers, for the sake of accuracy on OBP. With that restriction, the best pitcher is either Carlos Zambrano of the Cubs (16-8, 2.75 ERA, .257 OBP, Salary 450,000, 70 ABs, .646 OBP_Sal) or Jason Marquis of the Cardinals (15-7, 3.71 ERA, .297 OBP, Salary 525,000, 72 ABs, .895 OBP_Sal). Since this homework is supposed to focus on OBP, and since I'm a Cardinals fan, I'll go with Marquis on this one.